README.md
# FFlyShow

A Flutter plugin for iOS, Android for animation show content.

![The example app running in iOS](https://github.com/PreSwift/FFlyShow/blob/main/doc/ezgif-2-0643cf255401.gif)

[Feedback welcome](https://github.com/PreSwift/FFlyShow/issues) and
[Pull Requests](https://github.com/PreSwift/FFlyShow/pulls) are most welcome!

## Example

```dart
import 'package:ffly_show/ffly_show.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  var colors = [
    Colors.red,
    Colors.green,
    Colors.yellow,
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        padding: EdgeInsets.only(right: 20, top: 20),
        alignment: Alignment.topRight,
        decoration: BoxDecoration(
          color: Colors.white,
        ),
        clipBehavior: Clip.hardEdge,
        child: FFLyShow(
          itemWidth: 180,
          itemHeight: 200,
          children: getImageWidgets(),
        ),
      ),
    );
  }

  List<Widget> getImageWidgets() {
    List<Widget> imageWidgets = [];
    for (var i=0;i<colors.length;i++) {
      imageWidgets.add(
          Container(
            clipBehavior: Clip.hardEdge,
            decoration: BoxDecoration(
              borderRadius: BorderRadius.all(Radius.circular(5)),
            ),
            child: Container(
              color: colors[i],
            ),
          )
      );
    }
    return imageWidgets;
  }
}
```


CHANGELOG.md

## [0.0.2+1] - 2021-03-12.

* 添加README。

## [0.0.2] - 2021-02-05.

* 优化动画性能。

## [0.0.1] - 2021-01-28.

* 第一次上传.



example/ffly_show_example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:ffly_show_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/ffly_show.dart
library ffly_show;

import 'dart:async';
import 'dart:math';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

class FFlyMoveModel {

  FFlyMoveModel({
    @required this.widget,
    @required this.animationController,
    this.offsetAnimation,
    this.scaleAnimation,
    this.rotationZAnimation,
    this.offset : const Offset(0.0, 0.0),
    this.scale : 1.0,
    this.rotationZ : 0.0,
    this.valueNotifier,
  });

  final Widget widget;
  AnimationController animationController;
  Animation<Offset> offsetAnimation;
  Animation<double> scaleAnimation;
  Animation<double> rotationZAnimation;
  Offset offset;
  double scale;
  double rotationZ;

  // 计算角度用
  Point tempPoint;
  double tempRota;
  bool isBack;

  ValueNotifier<int> valueNotifier;

  void reloadRotationZ() {
    if (rotationZ == 0.0) {
      return;
    }
    while (rotationZ.abs() >= pi * 2) {
      rotationZ = (rotationZ.abs() / rotationZ) * (rotationZ.abs() - pi);
    }
    if (rotationZ.abs() > pi) {
      rotationZ = -(rotationZ.abs() / rotationZ) * (pi * 2 - rotationZ.abs());
    }
  }

}

class FFLyShow extends StatefulWidget {

  const FFLyShow({
    Key key,
    @required this.itemWidth,
    @required this.itemHeight,
    @required this.children,
    this.offsetH = 7,
    this.offsetV = 7,
  }) :  assert(itemWidth != null && itemHeight != null),
        assert(children != null && children.length >= 2),
        assert(offsetH != null && ((children.length - 1) * 2 * offsetH) < itemWidth),
        assert(offsetV != null && ((children.length - 1) * 2 * offsetV) < itemHeight),
        super(key: key);

  final double itemWidth;
  final double itemHeight;
  final double offsetH;
  final double offsetV;
  final List<Widget> children;

  @override
  State<StatefulWidget> createState() {
    return FFlyShowState();
  }
}

class FFlyShowState extends State<FFLyShow> with TickerProviderStateMixin {

  final _duration = Duration(milliseconds: 200);
  StreamController<int> _streamController;

  int itemCount;
  List<FFlyMoveModel> _currentChildren = [];
  List<FFlyMoveModel> _addChildren = [];
  List<FFlyMoveModel> _removeChildren = [];

  @override
  void initState() {
    _streamController = StreamController<int>();
    itemCount = widget.children.length;
    for (var i=0; i<itemCount; i++) {
      FFlyMoveModel model = FFlyMoveModel(widget: widget.children[i], animationController: AnimationController(duration: _duration, vsync: this));
      model.valueNotifier = ValueNotifier<int>(1);
      model.offset = Offset(0, widget.offsetV * (itemCount - 1 - i));
      model.scale = (widget.itemWidth - widget.offsetH * 2 * (itemCount - 1 - i)) / widget.itemWidth;
      model.animationController.addListener(() {
        model.valueNotifier.value = -model.valueNotifier.value;
      });
      model.animationController.addStatusListener((status) {
        if (status == AnimationStatus.completed) {
          model.animationController.reset();
          reloadLocation();
          model.valueNotifier.value = -model.valueNotifier.value;
        }
      });
      _currentChildren.add(model);
    }
    super.initState();
  }

  void reloadLocation() {
    for (int i=0;i<_currentChildren.length;i++) {
      FFlyMoveModel model = _currentChildren[i];
      model.offset = Offset(0, widget.offsetV *  (_currentChildren.length - 1 - i));
      model.scale = (widget.itemWidth - widget.offsetH * 2 * (_currentChildren.length - 1 - i)) / widget.itemWidth;
      model.rotationZ = 0.0;
    }
  }

  @override
  void dispose() {
    for (var i=0; i<_currentChildren.length; i++) {
      _currentChildren[i].animationController.stop();
      _currentChildren[i].animationController.dispose();
      // _currentChildren[i].valueNotifier.dispose();
    }
    for (var i=0; i<_addChildren.length; i++) {
      _addChildren[i].animationController.stop();
      _addChildren[i].animationController.dispose();
      // _addChildren[i].valueNotifier.dispose();
    }
    for (var i=0; i<_removeChildren.length; i++) {
      _removeChildren[i].animationController.stop();
      _removeChildren[i].animationController.dispose();
      // _removeChildren[i].valueNotifier.dispose();
    }
    _streamController.close();
    super.dispose();
  }

  // 计算三个点形成的角度
  double getAngle(Point pt1, Point pt2, Point pt0, double rate) {
    double maX = pt1.x - pt0.x;
    double maY = pt1.y - pt0.y;
    double mbX = pt2.x - pt0.x;
    double mbY = pt2.y - pt0.y;

    double x = (maX * mbX) + (maY * mbY);
    double y = (maX * mbY - mbX * maY);
    if (y == 0.0) {
      return 0.0;
    }
    return  y.abs() / y * acos(x / sqrt(x*x+y*y)) * rate;
  }

  double getDistanceRate(Point pt1, Point pt2) {
    if (pt1 == pt2) {
      return 1;
    }
    double maX = pt1.x - pt2.x;
    double maY = pt1.y - pt2.y;
    double dis = sqrt(maX * maX + maY * maY);
    double wDis = sqrt(widget.itemWidth * widget.itemWidth + widget.itemHeight * widget.itemHeight);
    return dis / wDis;
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onPanDown: (details) {
        FFlyMoveModel model = _currentChildren.last;
        model.tempPoint = Point(details.localPosition.dx, details.localPosition.dy);
        model.tempRota = getDistanceRate(model.tempPoint, Point(widget.itemWidth / 2, widget.itemHeight / 2));
      },
      onPanUpdate: (details) {
        FFlyMoveModel model = _currentChildren.last;
        if ((model.offset.dx + details.delta.dx).abs() < model.offset.dx.abs() || (model.offset.dy + details.delta.dy).abs() < model.offset.dy.abs()) {
          model.isBack = true;
        } else {
          model.isBack = false;
        }
        model.offset += details.delta;
        model.rotationZ += getAngle(model.tempPoint, Point(details.localPosition.dx, details.localPosition.dy), Point(widget.itemWidth / 2, widget.itemHeight / 2), model.tempRota);
        model.valueNotifier.value = -model.valueNotifier.value;
        model.tempPoint = Point(details.localPosition.dx, details.localPosition.dy);
      },
      onPanEnd: (details) {
        runAnimation(details);
      },
      child: Container(
        padding: EdgeInsets.only(bottom: (widget.children.length - 1) * widget.offsetV),
        child: SizedBox(
          width: widget.itemWidth,
          height: widget.itemHeight,
          child: StreamBuilder<int>(
              stream: _streamController.stream,
              builder: (context, snapshot) {
                return Stack(
                  alignment: Alignment.bottomCenter,
                  overflow: Overflow.clip,
                  children: getChildren(context),
                );
              }
          ),
        ),
      ),
    );
  }

  Widget buildTransformWidget(BuildContext context, FFlyMoveModel model) {
    return Transform.translate(
      offset: model.offset - (model.offsetAnimation?.value ?? Offset.zero),
      child: SizedBox(
        width: widget.itemWidth * (model.scale + (model.scaleAnimation?.value ?? 0.0)),
        height: widget.itemHeight * (model.scale + (model.scaleAnimation?.value ?? 0.0)),
        child: Transform.rotate(
          angle: model.rotationZ + (model.rotationZAnimation?.value ?? 0.0),
          child: model.widget,
        ),
      ),
    );
  }

  List<Widget> getChildren(BuildContext context) {
    List<Widget> children = [];
    for (var i=0; i<_addChildren.length; i++) {
      var model = _addChildren[i];
      children.add(
          ValueListenableBuilder(
            valueListenable: model.valueNotifier,
            builder: (context, snapshot, child) {
              return AnimatedBuilder(
                animation: model.animationController,
                child: buildTransformWidget(context, model),
                builder: (BuildContext context, Widget child) {
                  return Opacity(opacity: model.animationController.value, child: child,);
                },
              );
            },
          )
      );
    }
    for (var i=0; i<_currentChildren.length; i++) {
      var model = _currentChildren[i];
      children.add(ValueListenableBuilder(
        valueListenable: model.valueNotifier,
        builder: (context, snapshot, child) {
          return buildTransformWidget(context, model);
        },
      ));
    }
    for (var i=_removeChildren.length - 1; i>=0; i--) {
      var model = _removeChildren[i];
      children.add(ValueListenableBuilder(
        valueListenable: model.valueNotifier,
        builder: (context, snapshot, child) {
          return buildTransformWidget(context, model);
        },
      ));
    }
    return children;
  }

  void runAnimation(DragEndDetails details) {
    FFlyMoveModel model = _currentChildren.last;
    if ((details.velocity.pixelsPerSecond.dx.abs() < 50 && details.velocity.pixelsPerSecond.dy.abs() < 50) || (model.isBack)) {
      // 恢复
      model.offsetAnimation = Tween(begin: Offset.zero, end: model.offset).chain(CurveTween(curve: Curves.easeOutBack)).animate(model.animationController);
      model.reloadRotationZ();
      model.rotationZAnimation = Tween(begin: 0.0, end: -model.rotationZ).chain(CurveTween(curve: Curves.easeOutBack)).animate(model.animationController);
      model.animationController.forward();
    } else {
      // 移除最后一个
      FFlyMoveModel model = _currentChildren.removeLast();
      model.animationController.dispose();
      // model.valueNotifier.dispose();

      // 移除动画
      final transX = model.offset.dx, transY = model.offset.dy;
      final transV = sqrt(transX * transX + transY * transY);
      final queryWight =  MediaQuery.of(context).size.width, queryHeight =  MediaQuery.of(context).size.height;
      final queryV = sqrt(queryWight * queryWight + queryHeight * queryHeight);
      final v = queryV / transV;
      final dv = transV * v;
      final velocityX = details.velocity.pixelsPerSecond.dx.abs(), velocityY = details.velocity.pixelsPerSecond.dy.abs();
      final velocityV = sqrt(velocityX * velocityX + velocityY * velocityY);
      double seconds = dv / velocityV;
      seconds = max(seconds, 0.5);
      seconds = min(seconds, 2.0);
      FFlyMoveModel removeModel = FFlyMoveModel(widget: model.widget, offset: model.offset, animationController: AnimationController(duration: Duration(microseconds: (1000 * 1000 * seconds).toInt()), vsync: this));
      removeModel.valueNotifier = ValueNotifier<int>(1);
      removeModel.tempRota = model.tempRota;
      removeModel.tempPoint = model.tempPoint;
      removeModel.rotationZ = model.rotationZ;
      removeModel.reloadRotationZ();
      removeModel.animationController.addListener(() {
        removeModel.valueNotifier.value = -removeModel.valueNotifier.value;
      });
      removeModel.animationController.addStatusListener((status) {
        if (status == AnimationStatus.completed) {
          _removeChildren.remove(removeModel);
          removeModel.animationController.dispose();
          removeModel.valueNotifier.value = -removeModel.valueNotifier.value;
          // removeModel.valueNotifier.dispose();
        }
      });
      removeModel.offsetAnimation = Tween(begin: Offset.zero, end: model.offset - Offset(transX * v, transY * v)).animate(removeModel.animationController);
      double targetRota = min(model.rotationZ.abs() * 10, pi * 4);
      final velocityRota = velocityV * targetRota / queryV;
      final rotationT = (model.rotationZ.abs() / model.rotationZ) * velocityRota * seconds;
      removeModel.rotationZAnimation = Tween(begin: 0.0, end: rotationT).animate(removeModel.animationController);
      _removeChildren.insert(0, removeModel);
      removeModel.animationController?.forward();

      // 上升动画
      for (int i=0; i<_currentChildren.length; i++) {
        FFlyMoveModel currentModel = _currentChildren[i];
        currentModel.offsetAnimation = Tween(begin: Offset.zero, end: Offset(0, widget.offsetV)).animate(currentModel.animationController);
        currentModel.scaleAnimation = Tween(begin: 0.0, end: widget.offsetH * 2 / widget.itemWidth).animate(currentModel.animationController);
        currentModel.animationController.forward();
      }

      // 添加动画
      FFlyMoveModel addModel = FFlyMoveModel(widget: model.widget, animationController: AnimationController(duration: _duration, vsync: this));
      addModel.valueNotifier = ValueNotifier<int>(1);
      addModel.offset = Offset(0, widget.offsetV * (itemCount - 2));
      addModel.scale = (widget.itemWidth - widget.offsetH * 2 * (itemCount - 1)) / widget.itemWidth;
      addModel.animationController.addListener(() {
        addModel.valueNotifier.value = -addModel.valueNotifier.value;
      });
      addModel.animationController.addStatusListener((status) {
        if (status == AnimationStatus.completed) {
          _addChildren.remove(addModel);
          addModel.animationController.dispose();
          addModel.valueNotifier.value = -addModel.valueNotifier.value;
          // addModel.valueNotifier.dispose();

          FFlyMoveModel model = FFlyMoveModel(widget: addModel.widget, animationController: AnimationController(duration: _duration, vsync: this));
          model.valueNotifier = ValueNotifier<int>(1);
          model.offset = Offset(0, widget.offsetV * (itemCount - 1));
          model.scale = (widget.itemWidth - widget.offsetH * 2 * (itemCount - 1)) / widget.itemWidth;
          model.animationController.addListener(() {
            model.valueNotifier.value = -model.valueNotifier.value;
          });
          model.animationController.addStatusListener((status) {
            if (status == AnimationStatus.completed) {
              model.animationController.reset();
              reloadLocation();
              model.valueNotifier.value = -model.valueNotifier.value;
            }
          });
          _currentChildren.insert(0, model);
        }
      });
      addModel.offsetAnimation = Tween(begin: Offset.zero, end: -Offset(0, widget.offsetV)).animate(addModel.animationController);
      _addChildren.insert(0, addModel);
      addModel.animationController?.forward();

      _streamController.sink.add(1);
    }
  }
}


example/lib/main.dart
import 'package:ffly_show/ffly_show.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  var colors = [
    Colors.red,
    Colors.green,
    Colors.yellow,
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        padding: EdgeInsets.only(right: 20, top: 20),
        alignment: Alignment.topRight,
        decoration: BoxDecoration(
          color: Colors.white,
        ),
        clipBehavior: Clip.hardEdge,
        child: FFLyShow(
          itemWidth: 180,
          itemHeight: 200,
          children: getImageWidgets(),
        ),
      ),
    );
  }

  List<Widget> getImageWidgets() {
    List<Widget> imageWidgets = [];
    for (var i=0;i<colors.length;i++) {
      imageWidgets.add(
          Container(
            clipBehavior: Clip.hardEdge,
            decoration: BoxDecoration(
              borderRadius: BorderRadius.all(Radius.circular(5)),
            ),
            child: Container(
              color: colors[i],
            ),
          )
      );
    }
    return imageWidgets;
  }
}
